src/ExactDiagonalization/init_and_solvers.jl

```diff
@@ -0,0 +1,149 @@
+function CommonSolve.init( # no algorithm specified as positional argument
+    p::ExactDiagonalizationProblem;
+    kwargs...
+)
+    kw_nt = (; p.kw_nt..., kwargs...) # remove duplicates
+    if isdefined(kw_nt, :algorithm)
+        algorithm = kw_nt.algorithm
+        kw_nt = delete(kw_nt, (:algorithm,))
+    else
+        algorithm = LinearAlgebraEigen()
+    end
+
+    return init(p, algorithm; kw_nt...)
+end
+
+
+struct KrylovKitDirectEDSolver{A<:KrylovKitDirect,P,V<:PDVec,K<:NamedTuple}
+    algorithm::A
+    problem::P
+    v0::V
+    kw_nt::K
+end
+function Base.show(io::IO, s::KrylovKitDirectEDSolver)
+    io = IOContext(io, :compact => true)
+    print(io, "KrylovKitDirectEDSolver\n with algorithm $(s.algorithm) for h = ")
+    show(io, s.problem.h)
+    print(io, ",\n  v0 = ")
+    show(io, s.v0)
+    print(io, ",\n  kwargs = ")
+    show(io, s.kw_nt)
+    print(io, "\n)")
+end
+
+function CommonSolve.init(
+    p::ExactDiagonalizationProblem, algorithm::KrylovKitDirect;
+    kwargs...
+)
+    kw = (; p.kw_nt..., algorithm.kw_nt..., kwargs...) # remove duplicates
+    # set up the starting vector
+    vec = if isnothing(p.v0)
+        FrozenDVec([starting_address(p.h) => 1.0])
+    elseif p.v0 isa AbstractFockAddress
+        FrozenDVec([p.v0 => 1.0])
+    elseif p.v0 isa Union{
+        NTuple{<:Any,<:AbstractFockAddress},
+        AbstractVector{<:AbstractFockAddress},
+    }
+        FrozenDVec([addr => 1.0 for addr in p.v0])
+    elseif p.v0 isa FrozenDVec{<:AbstractFockAddress}
+        p.v0
+    else
+        throw(ArgumentError("Invalid starting vector in `ExactDiagonalizationProblem`."))
+    end
+    svec = PDVec(vec)
+
+    return KrylovKitDirectEDSolver(algorithm, p, svec, kw)
+end
+
+struct MatrixEDSolver{A,P,BSR<:BasisSetRep,V<:Union{Nothing,FrozenDVec},K<:NamedTuple}
+    algorithm::A
+    problem::P
+    basissetrep::BSR
+    v0::V
+    kw_nt::K # NamedTuple
+end
+
+function Base.show(io::IO, s::MatrixEDSolver)
+    io = IOContext(io, :compact => true)
+    b = s.basissetrep
+    print(io, "MatrixEDSolver\n  with algorithm = ")
+    show(io, s.algorithm)
+    print(io, "\n for h  = ")
+    show(io, s.problem.h)
+    print(io, ":\n  ")
+    show(io, MIME"text/plain"(), b.sm)
+    print(io, ",\n  v0 = ")
+    show(io, s.v0)
+    print(io, ",\n  kwargs = ")
+    show(io, NamedTuple(s.kw_nt))
+    print(io, "\n)")
+end
+
+# init with matrix-based algorithms
+function CommonSolve.init(
+    p::ExactDiagonalizationProblem, algorithm::ALG;
+    kwargs...
+) where {ALG<:Union{KrylovKitMatrix,LinearAlgebraEigen,ArpackEigs,LOBPCG}}
+    !ishermitian(p.h) && algorithm isa LOBPCG &&
+        @warn "LOBPCG() is not suitable for non-hermitian matrices."
+
+    # set keyword arguments for BasisSetRep
+    kw = (; p.kw_nt..., algorithm.kw_nt..., kwargs...) # remove duplicates
+    if isdefined(kw, :sizelim)
+        sizelim = kw.sizelim
+    elseif algorithm isa LinearAlgebraEigen
+        sizelim = 10^4 # default for dense matrices
```

Comment:
```suggestion
        sizelim = 10^5 # default for dense matrices
```
It would be good to estimate how much memory different algorithms use and decide what we consider reasonable.